File: app/src/main/java/com/amyllykoski/asyncseq/impl/ServiceHandler.java
package com.amyllykoski.asyncseq.impl;

import android.os.Handler;
import android.os.Looper;
import android.os.Message;

import com.amyllykoski.asyncseq.api.RestCallback;

public class ServiceHandler<T> extends Handler {

  static final int MSG_OK = 1;
  static final int MSG_NOK = 2;

  private RestCallback<T> restCallback;

  public ServiceHandler(Looper looper, RestCallback<T> restCallback) {
    super(looper);
    if (restCallback == null) {
      throw new IllegalArgumentException("RestCallback cannot be null.");
    }
    this.restCallback = restCallback;
  }

  @Override
  @SuppressWarnings("unchecked")
  public void handleMessage(Message msg) {
    switch (msg.what) {
      case MSG_OK:
      if(msg.obj != null) restCallback.onResponse((T) msg.obj);
        break;
      default:
        String error = msg.obj != null ? msg.obj.toString() : "?";
        restCallback.onFailure(error);
        break;
    }
  }
}

File: app/src/main/java/com/amyllykoski/asyncseq/impl/ServiceHandler.kt
package com.amyllykoski.asyncseq.impl

import android.os.Handler
import android.os.Looper
import android.os.Message
import arrow.core.Either
import com.amyllykoski.asyncseq.api.MyError.AnotherError
import com.amyllykoski.asyncseq.api.MyError.ServerError

import com.amyllykoski.asyncseq.api.RestCallbackWithEither

class ServiceHandler<T>(
    looper: Looper,
    private val restCallback: RestCallbackWithEither<T>) : Handler(looper) {

  @Suppress("UNCHECKED_CAST")
  override fun handleMessage(msg: Message) = when (msg.what) {
    MSG_OK ->
      restCallback.onResponse(Either.Right(msg.obj as T))
    MSG_NOK ->
      restCallback.onResponse(Either.Left(ServerError))
    else ->
      restCallback.onResponse(Either.Left(AnotherError(msg.obj as String)))
  }

  companion object {
    const val MSG_OK = 1
    const val MSG_NOK = 2
  }
}
